src/tools/rbd_mirror/image_replayer/CreateLocalImageRequest.cc

```diff
@@ -0,0 +1,311 @@
+  int r = m_local_io_ctx.aio_operate(RBD_MIRRORING, rados_completion, &op);
+  assert(r == 0);
+  rados_completion->release();
+}
+
+template<typename I>
+Context* CreateLocalImageRequest<I>::handle_create_local_image_checkpoint_remove(int *result) {
+  dout(20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    derr << ": failed to remove chekpoint post image create failure: "
+         << cpp_strerror(*result) << dendl;
+  }
+
+  *result = m_ret_val;
+  return m_on_finish;
+}
+
+template<typename I>
+void CreateLocalImageRequest<I>::create_local_image_checkpoint_end() {
+  dout(20) << dendl;
+
+  m_mirror_image.state = cls::rbd::MIRROR_IMAGE_STATE_ENABLED;
+
+  librados::ObjectWriteOperation op;
+  librbd::cls_client::mirror_image_set(&op, m_curr_image_id, m_mirror_image);
+
+  using klass = CreateLocalImageRequest<I>;
+  librados::AioCompletion *rados_completion = create_rados_ack_callback
+    <klass, &klass::handle_create_local_image_checkpoint_end>(this);
+  int r = m_local_io_ctx.aio_operate(RBD_MIRRORING, rados_completion, &op);
+  assert(r == 0);
+  rados_completion->release();
+}
+
+template<typename I>
+Context* CreateLocalImageRequest<I>::handle_create_local_image_checkpoint_end(int *result) {
+  dout(20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    // It gets bad here -- the image was successfully created but
+    // the checkpoint could not be marked. This tricks subsequent
+    // bootstraps (for this image) into assuming that image creation
+    // was interrupted in the last run. So, it's better to rollback
+    // and remove the image.
+    m_ret_val = *result;
+    remove_local_image_done();
+    return nullptr;
+  }
+
+  *m_local_image_id = m_curr_image_id;
+
+  send_notify_watcher();
+  return nullptr;
+}
+
+template<typename I>
+void CreateLocalImageRequest<I>::send_notify_watcher() {
+  dout(20) << dendl;
+
+  using klass = CreateLocalImageRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_notify_watcher>(this);
+
+  librbd::MirroringWatcher<>::notify_image_updated(
+    m_local_io_ctx,cls::rbd::MIRROR_IMAGE_STATE_ENABLED, m_curr_image_id,
+    m_global_image_id, ctx);
+}
+
+template<typename I>
+Context* CreateLocalImageRequest<I>::handle_notify_watcher(int *result) {
+  dout(20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    derr << ": failed to send update notification: " << cpp_strerror(*result) << dendl;
+  }
+
+  *result = 0;
+  return m_on_finish;
+}
+
+template<typename I>
+void CreateLocalImageRequest<I>::remove_local_image_done() {
+  dout(20) << dendl;
+
+  using klass = CreateLocalImageRequest<I>;
+  Context *ctx = create_context_callback
+    <klass, &klass::handle_remove_local_image_done>(this);
+
+  librbd::image::RemoveRequest<I> *req = librbd::image::RemoveRequest<I>::create(
+    m_local_io_ctx, m_local_image_name, m_curr_image_id, true, m_no_op_prog_ctx,
+    m_op_work_queue, ctx);
+  req->send();
```

Comment:
@vshankar it could be a stupid question, but what's the benifits of an async-ed removing here? I found the most of our operations are being re-implemented as a async-style, I am sure there should be some reasons.  could you explain the advantage of it compared with the blocking way. thanx a lot.